### ISSUE TITLE
Multiple component updates with invalid states when splitting list item

### ISSUE THREAD

**Issue by mchowning** [wordpress-mobile/gutenberg-mobile#2204]
**Describe the bug**
On Android, splitting a list item appears to cause multiple updates to the text. This causes a number of issues including (a) sometimes ending up in an unexpected state after splitting list items (see Step 5 in the steps to reproduce) and (b) it can take multiple undo actions to "undo" the splitting of a list item (see Steps 6, 8, 10, 12, and 14).

When I tested this on production (a) did not occur, but (b) still did. I suspect that (a) is due to a race condition.

**To Reproduce**
1. Create a new list
2. Type two letters in the list item ("gg")
3. Put your cursor between the two letters ("g|g")
4. Tap Enter
5. On a release build, I see the list properly split in two (two list items, each with a single "g". On a debug build, however, I briefly see a second list item with the "g", but then the "g" in the second list item disappears, so the result is that the first list item properly has a single "g", but the second list item is empty instead of having a single "g" as it should.
6. Tap undo
7. Observe that the "g" in the second list item reappeared, but the bullet for the second list item is now missing
8. Tap undo a second time
9. Observe that the bullet for the second list item returns, so now the second list item appears both with the bullet and single "g" letter. This is how it should have looked immediately after step 4.
10. Tap undo a third time.
11. Observe that the "g" in the second list item disappears, but the second list item remains (now it is empty), and the cursor has returned to the first list item.
12. Tap undo a fourth time
13. Observe that the second list item disappears, and the first (and only) list item is unchanged (only has a single "g")
14. Tap undo a fifth time
15. Observe the first and only list item now has the two letters with the cursor between them, like it was in Step 3.

**Expected behavior**
1. Tapping Enter to split a list item should always result in the text after the cursor (if any) being present in the second list item); and
2. Tapping "undo" one time immediately after splitting a list item should undo the splitting of the list item.

**Screenshots**

This gif follows the steps to reproduce outlined above.

![list_undo mp4](https://user-images.githubusercontent.com/4656348/80750949-f7129700-8af6-11ea-9d6d-4a82fd87cb12.gif)

**Smartphone (please complete the following information):**
 - Device: Pixel, Pixel 3
 - OS: Android 10
 - Version alpha-219

**Comment by mchowning** [comment]
Possibly related to #65

**Comment by hypest** [comment]
I see the expected behavior (clear split, one-undo-returns to joint line) on (production) WPAndroid 15.0-rc-3 running on Pixel 2XL Android 10. Same on the demo app (against Metro) on 7495fbf3, same device/OS. Except for the slowness on the Metro case, I see no visual artifacts, broken/empty lines or similar.

Is the issue still happening for you @mchowning ?

**Comment by mchowning** [comment]
> I see the expected behavior (clear split, one-undo-returns to joint line) on (production) WPAndroid 15.0-rc-3

Just did a quick test using WPAndroid 15.0-rc-3 on a Pixel 3 Android 10, and I had to press undo 5 times to "undo" splitting a list item. I'm curious why I'm seeing the issue and you're not @hypest . 🤔  

![undo_split mp4](https://user-images.githubusercontent.com/4656348/84508780-cfcfed80-ac90-11ea-92f4-b69c4e051cd9.gif)

**Comment by hypest** [comment]
For reference, here's a GIF from my run:

![list-split-works-undo-too](https://user-images.githubusercontent.com/1032913/84509313-2eff2380-accc-11ea-9300-af5e14316cd4.gif)

**Comment by mchowning** [comment]
Ok, did some more testing, and apparently it makes a difference whether you move the cursor to the middle of the word by tapping or scrolling with the spacebar!?!?!?! 😱 

tapping | scrolling with spacebar
---- | ----
<img width="250" src="https://user-images.githubusercontent.com/4656348/84511360-65b94780-ac94-11ea-886d-5d970330c725.gif" /> | <img width="250" src="https://user-images.githubusercontent.com/4656348/84511374-69e56500-ac94-11ea-9a83-87591e17508d.gif" />

![no_sense](https://media.giphy.com/media/ycN7DLubUzQgFYhZAj/giphy.gif)
<img src="https://media.giphy.com/media/ycN7DLubUzQgFYhZAj/giphy.gif" />

**Comment by hypest** [comment]
Aha, interesting. If the "tap to set cursor" works, I'm thinking of lowering the priority on this ticket since I assume tapping to set cursor is the most common flow?

**Comment by mchowning** [comment]
> I'm thinking of lowering the priority on this ticket since I assume tapping to set cursor is the most common flow?

I agree. Went ahead and lowered this to Medium, but I'm not opposed to making it Low if you think that's more appropriate.

> can't repro on my side by using the spacebar-sliding method

So weird! I can't deny that it seems more likely I have something going on with my end compared to the idea that how you set the cursor really does make a difference here, but I've reproduced this on both the release 15.0-rc3 apk from github and the one distributed on the Play Store beta channel on two different Pixel devices (1 and 3). I'm not sure what else to try. If someone else tries this and can't recreate it we should probably just close this and we can reopen it if someone else encounters it.

**Comment by hypest** [comment]
> If someone else tries this and can't recreate it we should probably just close this and we can reopen it if someone else encounters it.

Let's give this a shot. @cameronvoell , can you try to reproduce this issue when you get the chance? Thanks!

**Comment by cameronvoell** [comment]
> Let's give this a shot. @cameronvoell , can you try to reproduce this issue when you get the chance? Thanks!

I am able to reproduce the multiple undo behavior ((b) from description ) follow the steps precisely from the description with scrolling spacebar, but not the "unexpected state" ((a) from description).

I can only reproduce (b) when I do the "scrolling with the spacebar". Tapping or moving the cursor with the arrow keys on my tablet do not reproduce the issue. It would be interesting to see why the spacebar scroll is different, but I agree this is less priority since the tap to move cursor works fine, and the scrolling with spacebar seems to only cause the issue if I follow the description steps exactly, and does not create extra undo's other times.

**Comment by guarani** [comment]
In https://github.com/wordpress-mobile/gutenberg-mobile/issues/2373#issuecomment-659798095 I learned that Aztec on Android doesn't handle input from Gboard well, especially in relation to Enter and Backspace keyboard events.

I've been able to reproduce this using Gboard, but can't reproduce using the Samsung keyboard. 

Here's a summary of my tests:

#### Gutenberg demo app on a Samsung Galaxy S10 running Android 10

When placing the cursor by _tapping_, I could reproduce on the second attempt.

<details>
<summary>See gif</summary>
<img width="250" src="https://user-images.githubusercontent.com/1898325/87991865-edc11700-cab4-11ea-908a-c692c2091384.gif">
</details>

When placing the cursor by scrolling the spacebar, I could often reproduce on the first go. I don't have a gif of this.

Also, after hitting Enter to create the second list item, the app sometimes got in a loop.

<details>
<summary>See gif</summary>
<img width="250" src="https://user-images.githubusercontent.com/1898325/87987245-52c43f00-caac-11ea-85a4-f895a4a84b7a.gif">
</details>


#### WPAndroid 15.2.1 on a Samsung Galaxy S10 running Android 10

Similar results to Gutenberg demo app tests above.

<details>
<summary>See gif</summary>
<img width="250" src="https://user-images.githubusercontent.com/1898325/87992190-bc951680-cab5-11ea-9a8e-afb7b92a90ac.gif">
</details>

**Comment by guarani** [comment]
I couldn't reproduce this on a Samsung S10 with Android 11, in both WPAndroid 18.2-rc-3 and Gutenberg demo app https://github.com/WordPress/gutenberg/commit/b815cb53cec99fd3589242fdd6678601d394fb31.

The things that have changed since https://github.com/wordpress-mobile/gutenberg-mobile/issues/2204#issuecomment-661380621 where I was able to reproduce are:
- Newer Gutenberg/WPAndroid versions
- Newer Android version (previously 10, now 11)

These are the scenarios I tested when trying to reproduce the issue:

**WPAndroid 18.2-rc-3**

- Split list item by manually placing cursor (using gboard)
- Split list item by scrolling the spacebar (using Samsung keyboard)

**Gutenberg demo app https://github.com/WordPress/gutenberg/commit/b815cb53cec99fd3589242fdd6678601d394fb31**

The demo app is very much slower than WPAndroid, but I still can't reproduce this issue.

- Split list item by manually placing cursor (using gboard)
- Split list item by scrolling the spacebar (using Samsung keyboard)

Since it's been over a year since this issue was last commented on, I think the best option would be to close it, especially if @mchowning or someone else can confirm it's no longer reproducible. Alternatively, we could give it the low priority label instead.
